client/python/gradio_client/client.py

```diff
@@ -1059,68 +1063,51 @@ def _predict(*data) -> tuple:
-        for i, fs in enumerate(file_paths):
-            if not isinstance(fs, list):
-                fs = [fs]
-            for f in fs:
-                files.append(("files", (Path(f).name, open(f, "rb"))))  # noqa: SIM115
-                indices.append(i)
-        r = httpx.post(
-            self.client.upload_url,
-            headers=self.client.headers,
-            cookies=self.client.cookies,
-            files=files,
-        )
-        if r.status_code != 200:
-            uploaded = file_paths
-        else:
-            uploaded = []
-            result = r.json()
-            for i, fs in enumerate(file_paths):
-                if isinstance(fs, list):
-                    output = [o for ix, o in enumerate(result) if indices[ix] == i]
-                    res = [
-                        {
-                            "path": o,
-                            "orig_name": Path(f).name,
-                        }
-                        for f, o in zip(fs, output)
-                    ]
-                else:
-                    o = next(o for ix, o in enumerate(result) if indices[ix] == i)
-                    res = {
-                        "path": o,
-                        "orig_name": Path(fs).name,
-                    }
-                uploaded.append(res)
-        return uploaded
-
-    def insert_state(self, *data) -> tuple:
+    def insert_empty_state(self, *data) -> tuple:
         data = list(data)
         for i, input_component_type in enumerate(self.input_component_types):
             if input_component_type.is_state:
                 data.insert(i, None)
         return tuple(data)
 
+    def process_input_files(self, *data) -> tuple:
+        data_ = []
+        for i, d in enumerate(data):
+            if self.client.upload_files and self.input_component_types[i].value_is_file:
+                d = utils.traverse(
+                    d,
+                    self._upload_file,
+                    lambda f: utils.is_filepath(f)
+                    or utils.is_file_obj_with_meta(f)
+                    or utils.is_http_url_like(f),
+                )
+            elif not self.client.upload_files:
+                d = utils.traverse(d, self._upload_file, utils.is_file_obj_with_meta)
+            data_.append(d)
+        return tuple(data_)
+
+    def process_predictions(self, *predictions):
+        # If self.download_file is True, we assume that that the user is using the Client directly (as opposed
+        # within gr.load) and therefore, download any files generated by the server and skip values for
+        # components that the user likely does not want to see (e.g. gr.State, gr.Tab).
+        if self.client.download_files:
+            predictions = self.download_files(*predictions)
+            predictions = self.remove_skipped_components(*predictions)
+        predictions = self.reduce_singleton_output(*predictions)
+        return predictions
+
+    def download_files(self, *data) -> tuple:
```

Comment:
Hmm we could do this, but there's still a chance of false positives. Let's say that your function returns a json that is of this format: {"path": "/url/to/image"}, which is not *that* unlikely

This will fail because `is_file_obj_new(data)` will parse the json dict {"path": "/url/to/image"} as a file and try to download it.